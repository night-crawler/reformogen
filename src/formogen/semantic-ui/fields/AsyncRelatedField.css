.AsyncRelatedField div[role="option"] {
    z-index: 10000000;
}