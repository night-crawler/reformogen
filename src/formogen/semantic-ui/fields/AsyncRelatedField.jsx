import PropTypes from 'prop-types';
import React from 'react';
import { isEmpty } from 'lodash';
import { Form } from 'semantic-ui-react';
import { SelectBase } from 'react-select';

import propTypes from '../../fieldPropTypes';
import { FieldLabel } from '../FieldLabel';
import { ErrorsList } from '../ErrorsList';


import './AsyncRelatedField.css';

export class AsyncRelatedField extends React.Component {
  static propTypes = {
    ...propTypes,
    getOptionLabel: PropTypes.func,
    getOptionValue: PropTypes.func,
    isMulti: PropTypes.bool,
  };

  static defaultProps = {
    isMulti: false,
    loadOptions: payload => // eslint-disable-next-line
      console.info(`AsyncRelatedField.loadOptions(${JSON.stringify(payload)})`),
  
    getOptionLabel: ({ name, title }) => name || title,
    getOptionValue: ({ id }) => id,
  };

  constructor(props) {
    super(props);

    this.storedCallback = () => {};
    this.storedInputText = '';
    this.state = {
      openCounter: 0,
      menuIsOpen: false,
      isLoading: false,
      search: '',
    };
  }

  handleChange = value => this.props.onChange(null, { 
    name: this.props.name, 
    value: value
  });

  render() {
    return (
      <Form.Field
        className='AsyncRelatedField'
        required={ this.props.required }
        disabled={ !this.props.editable }
        width={ this.props.displayOptions.width }
        error={ !isEmpty(this.props.errors) }
      >
        <FieldLabel { ...this.props } />
        <SelectBase
          ref={ self => this.select = self }
          inputId={ `AsyncRelatedField-${this.props.formId}-${this.props.name}` }
          options={ this.props.options }
          
          inputValue={ this.state.search }
          placeholder={ this.props.placeholder }
          value={ this.props.value }

          menuIsOpen={ this.state.menuIsOpen }
          isClearable={ this.props.editable }
          isDisabled={ !this.props.editable }
          isLoading={ this.state.isLoading }
          isRtl={ this.props.isRtl }
          isMulti={ this.props.isMulti }
          
          hideSelectedOptions={ false }

          getOptionLabel={ this.props.getOptionLabel }
          getOptionValue={ this.props.getOptionValue }
  
          onChange={ this.handleChange } 
          onMenuClose={ this.handleMenuClose }
          onMenuOpen={ this.handleMenuOpen }
          onInputChange={ this.handleInputChange }
          onMenuScrollToBottom={ this.handleMenuScrollToBottom }
        />
        { !this.props.helpTextOnHover
          ? <span className='help-text'>{ this.props.help_text }</span>
          : ''
        }
        <ErrorsList messages={ this.props.errors } />
      </Form.Field>
    );
  }

  handleMenuClose = () => {
    // eslint-disable-next-line
    this.setState({
      search: '',
      menuIsOpen: false,
      isLoading: false,
    });
  }
  handleMenuOpen = async () => {
    // eslint-disable-next-line
    await this.setState(
      { menuIsOpen: true, openCounter: this.state.openCounter + 1 },
      async () => {
        // ! some magic with constants
        // When users trigger handleMenuOpen they always have the inputText cleared, so it's ok
        // to call the new search with '' query.
        // When user opens the menu we must trigger loadOptions once only to fix this strange react-select
        // behaviour when it just does not trigger handleMenuScrollToBottom on the first page.
        // Any further pagination logic must be covered in handleMenuScrollToBottom separately.
        if (this.state.openCounter === 1)
          await this.loadOptions();
      }
    );
  }
  handleInputChange = async search => {
    if (search === this.state.search)
      return;
    // eslint-disable-next-line
    await this.setState(
      { search },
      async () => await this.loadOptions()
    );
  }
  handleMenuScrollToBottom = async () => {
    await this.loadOptions();
  }
  async loadOptions() {
    const { loadOptions } = this.props;

    if (this.state.isLoading) 
      return;
    
    // eslint-disable-next-line
    await this.setState({ isLoading: true });

    await loadOptions({ 
      formId: this.props.formId,  
      searchText: this.state.search, 
      fieldName: this.props.name,
      url: this.props.data,
      callback: async () => { 
        // eslint-disable-next-line
        await this.setState({ isLoading: false });
      },
    });
  }

  componentDidMount = async () => {
    // we need this bullshit here because react-select does not trigger 
    // onMenuScrollToBottom for the FIRST time it's being rendered
    await this.loadOptions();
  }
}
