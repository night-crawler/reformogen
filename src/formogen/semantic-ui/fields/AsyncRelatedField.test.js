import { AsyncRelatedField } from './AsyncRelatedField';


it('<AsyncRelatedField />', () => {
  const wrapper = shallow(
    <AsyncRelatedField
      type='AsyncRelatedField'
      name='AsyncRelatedField: name'
      verbose_name='AsyncRelatedField: verbose_name'
      help_text='AsyncRelatedField: help_text'
      layoutOpts={ { width: 1 } }
    />
  );
  expect(wrapper).toMatchSnapshot();
});
