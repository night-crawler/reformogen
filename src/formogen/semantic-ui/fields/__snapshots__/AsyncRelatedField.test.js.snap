// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AsyncRelatedField /> 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <AsyncRelatedField
    getOptionLabel={[Function]}
    getOptionValue={[Function]}
    help_text="AsyncRelatedField: help_text"
    isMulti={false}
    layoutOpts={
      Object {
        "width": 1,
      }
    }
    loadOptions={[Function]}
    name="AsyncRelatedField: name"
    type="AsyncRelatedField"
    verbose_name="AsyncRelatedField: verbose_name"
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "function",
    "props": Object {
      "children": Array [
        <FieldLabel
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          help_text="AsyncRelatedField: help_text"
          isMulti={false}
          layoutOpts={
            Object {
              "width": 1,
            }
          }
          loadOptions={[Function]}
          name="AsyncRelatedField: name"
          type="AsyncRelatedField"
          verbose_name="AsyncRelatedField: verbose_name"
        />,
        <Select
          backspaceRemovesValue={true}
          blurInputOnSelect={true}
          captureMenuScroll={false}
          closeMenuOnScroll={false}
          closeMenuOnSelect={true}
          components={Object {}}
          controlShouldRenderValue={true}
          escapeClearsValue={false}
          filterOption={[Function]}
          formatGroupLabel={[Function]}
          getOptionLabel={[Function]}
          getOptionValue={[Function]}
          hideSelectedOptions={false}
          inputId="AsyncRelatedField-undefined-AsyncRelatedField: name"
          inputValue=""
          isDisabled={true}
          isLoading={true}
          isMulti={false}
          isOptionDisabled={[Function]}
          isRtl={false}
          isSearchable={true}
          loadingMessage={[Function]}
          maxMenuHeight={300}
          menuIsOpen={false}
          menuPlacement="bottom"
          menuPosition="absolute"
          menuShouldBlockScroll={false}
          menuShouldScrollIntoView={true}
          minMenuHeight={140}
          noOptionsMessage={[Function]}
          onChange={[Function]}
          onInputChange={[Function]}
          onMenuClose={[Function]}
          onMenuOpen={[Function]}
          onMenuScrollToBottom={[Function]}
          openMenuOnClick={true}
          openMenuOnFocus={false}
          options={Array []}
          pageSize={5}
          placeholder="Select..."
          screenReaderStatus={[Function]}
          styles={Object {}}
          tabIndex="0"
          tabSelectsValue={true}
        />,
        <span
          className="help-text"
        >
          AsyncRelatedField: help_text
        </span>,
        <ErrorsList
          color="red"
          header="Errors"
        />,
      ],
      "className": "AsyncRelatedField",
      "disabled": true,
      "error": false,
      "required": undefined,
      "width": 1,
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "getOptionLabel": [Function],
          "getOptionValue": [Function],
          "help_text": "AsyncRelatedField: help_text",
          "isMulti": false,
          "layoutOpts": Object {
            "width": 1,
          },
          "loadOptions": [Function],
          "name": "AsyncRelatedField: name",
          "type": "AsyncRelatedField",
          "verbose_name": "AsyncRelatedField: verbose_name",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "class",
        "props": Object {
          "backspaceRemovesValue": true,
          "blurInputOnSelect": true,
          "captureMenuScroll": false,
          "closeMenuOnScroll": false,
          "closeMenuOnSelect": true,
          "components": Object {},
          "controlShouldRenderValue": true,
          "escapeClearsValue": false,
          "filterOption": [Function],
          "formatGroupLabel": [Function],
          "getOptionLabel": [Function],
          "getOptionValue": [Function],
          "hideSelectedOptions": false,
          "inputId": "AsyncRelatedField-undefined-AsyncRelatedField: name",
          "inputValue": "",
          "isClearable": undefined,
          "isDisabled": true,
          "isLoading": true,
          "isMulti": false,
          "isOptionDisabled": [Function],
          "isRtl": false,
          "isSearchable": true,
          "loadingMessage": [Function],
          "maxMenuHeight": 300,
          "menuIsOpen": false,
          "menuPlacement": "bottom",
          "menuPosition": "absolute",
          "menuShouldBlockScroll": false,
          "menuShouldScrollIntoView": true,
          "minMenuHeight": 140,
          "noOptionsMessage": [Function],
          "onChange": [Function],
          "onInputChange": [Function],
          "onMenuClose": [Function],
          "onMenuOpen": [Function],
          "onMenuScrollToBottom": [Function],
          "openMenuOnClick": true,
          "openMenuOnFocus": false,
          "options": Array [],
          "pageSize": 5,
          "placeholder": "Select...",
          "screenReaderStatus": [Function],
          "styles": Object {},
          "tabIndex": "0",
          "tabSelectsValue": true,
          "value": undefined,
        },
        "ref": [Function],
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "AsyncRelatedField: help_text",
          "className": "help-text",
        },
        "ref": null,
        "rendered": "AsyncRelatedField: help_text",
        "type": "span",
      },
      Object {
        "instance": null,
        "key": undefined,
        "nodeType": "function",
        "props": Object {
          "color": "red",
          "header": "Errors",
          "messages": undefined,
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": [Function],
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "function",
      "props": Object {
        "children": Array [
          <FieldLabel
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            help_text="AsyncRelatedField: help_text"
            isMulti={false}
            layoutOpts={
              Object {
                "width": 1,
              }
            }
            loadOptions={[Function]}
            name="AsyncRelatedField: name"
            type="AsyncRelatedField"
            verbose_name="AsyncRelatedField: verbose_name"
          />,
          <Select
            backspaceRemovesValue={true}
            blurInputOnSelect={true}
            captureMenuScroll={false}
            closeMenuOnScroll={false}
            closeMenuOnSelect={true}
            components={Object {}}
            controlShouldRenderValue={true}
            escapeClearsValue={false}
            filterOption={[Function]}
            formatGroupLabel={[Function]}
            getOptionLabel={[Function]}
            getOptionValue={[Function]}
            hideSelectedOptions={false}
            inputId="AsyncRelatedField-undefined-AsyncRelatedField: name"
            inputValue=""
            isDisabled={true}
            isLoading={true}
            isMulti={false}
            isOptionDisabled={[Function]}
            isRtl={false}
            isSearchable={true}
            loadingMessage={[Function]}
            maxMenuHeight={300}
            menuIsOpen={false}
            menuPlacement="bottom"
            menuPosition="absolute"
            menuShouldBlockScroll={false}
            menuShouldScrollIntoView={true}
            minMenuHeight={140}
            noOptionsMessage={[Function]}
            onChange={[Function]}
            onInputChange={[Function]}
            onMenuClose={[Function]}
            onMenuOpen={[Function]}
            onMenuScrollToBottom={[Function]}
            openMenuOnClick={true}
            openMenuOnFocus={false}
            options={Array []}
            pageSize={5}
            placeholder="Select..."
            screenReaderStatus={[Function]}
            styles={Object {}}
            tabIndex="0"
            tabSelectsValue={true}
          />,
          <span
            className="help-text"
          >
            AsyncRelatedField: help_text
          </span>,
          <ErrorsList
            color="red"
            header="Errors"
          />,
        ],
        "className": "AsyncRelatedField",
        "disabled": true,
        "error": false,
        "required": undefined,
        "width": 1,
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "getOptionLabel": [Function],
            "getOptionValue": [Function],
            "help_text": "AsyncRelatedField: help_text",
            "isMulti": false,
            "layoutOpts": Object {
              "width": 1,
            },
            "loadOptions": [Function],
            "name": "AsyncRelatedField: name",
            "type": "AsyncRelatedField",
            "verbose_name": "AsyncRelatedField: verbose_name",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "class",
          "props": Object {
            "backspaceRemovesValue": true,
            "blurInputOnSelect": true,
            "captureMenuScroll": false,
            "closeMenuOnScroll": false,
            "closeMenuOnSelect": true,
            "components": Object {},
            "controlShouldRenderValue": true,
            "escapeClearsValue": false,
            "filterOption": [Function],
            "formatGroupLabel": [Function],
            "getOptionLabel": [Function],
            "getOptionValue": [Function],
            "hideSelectedOptions": false,
            "inputId": "AsyncRelatedField-undefined-AsyncRelatedField: name",
            "inputValue": "",
            "isClearable": undefined,
            "isDisabled": true,
            "isLoading": true,
            "isMulti": false,
            "isOptionDisabled": [Function],
            "isRtl": false,
            "isSearchable": true,
            "loadingMessage": [Function],
            "maxMenuHeight": 300,
            "menuIsOpen": false,
            "menuPlacement": "bottom",
            "menuPosition": "absolute",
            "menuShouldBlockScroll": false,
            "menuShouldScrollIntoView": true,
            "minMenuHeight": 140,
            "noOptionsMessage": [Function],
            "onChange": [Function],
            "onInputChange": [Function],
            "onMenuClose": [Function],
            "onMenuOpen": [Function],
            "onMenuScrollToBottom": [Function],
            "openMenuOnClick": true,
            "openMenuOnFocus": false,
            "options": Array [],
            "pageSize": 5,
            "placeholder": "Select...",
            "screenReaderStatus": [Function],
            "styles": Object {},
            "tabIndex": "0",
            "tabSelectsValue": true,
            "value": undefined,
          },
          "ref": [Function],
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "host",
          "props": Object {
            "children": "AsyncRelatedField: help_text",
            "className": "help-text",
          },
          "ref": null,
          "rendered": "AsyncRelatedField: help_text",
          "type": "span",
        },
        Object {
          "instance": null,
          "key": undefined,
          "nodeType": "function",
          "props": Object {
            "color": "red",
            "header": "Errors",
            "messages": undefined,
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": [Function],
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
  },
}
`;
